### ISSUE TITLE
PPO grad_norm is 0.0

### ISSUE THREAD

**Issue by faker52** [huggingface/trl#3961]
I want to visualize the gradient of PPO training through the following code, but the gradients I get are all 0. Why?

<img width="517" height="478" alt="Image" src="https://github.com/user-attachments/assets/f2a41745-0ab1-46c7-8997-8489328b2bf7" />

<img width="390" height="280" alt="Image" src="https://github.com/user-attachments/assets/e46da24b-fb74-4e70-94f9-4c655b2e0ae7" />

**Comment by LeonEricsson** [comment]
Use [clip_grad_norm](https://huggingface.co/docs/accelerate/en/package_reference/accelerator#accelerate.Accelerator.clip_grad_norm_) instead (with a high max norm to avoid clipping), it returns the unclipped gradient norm. Is that also 0?

**Comment by qgallouedec** [comment]
Hey @faker52, sorry for the slow reply, we're catching up on the backlog.

The behavior you're seeing isn't a bug: PPO calls `optimizer.zero_grad()` at the end of every microbatch update (see https://github.com/huggingface/trl/blob/51c6d3ca31fb4cc80ff719c0844bbdfcd0feeefb/trl/experimental/ppo/ppo_trainer.py#L856-L858), so any `param.grad` you read after the step has been cleared to zero

**Comment by qgallouedec** [comment]
Hey @faker52, sorry for the slow reply, we're catching up on the backlog.

If you want the actual gradient norm, do as @LeonEricsson suggested and call `accelerator.clip_grad_norm_` with a large `max_norm` (it returns the unclipped norm). Note that the PPO trainer has also moved to `trl.experimental.ppo`.
